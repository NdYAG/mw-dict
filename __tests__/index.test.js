const { CollegiateDictionary, WordNotFoundError } = require('../')

const dict = new CollegiateDictionary('')

describe('after fetching xml document from dictionaryapi.com', () => {
  it('parses <ew>, <fl>, and <def>', () => {
    return dict.lookup('TEST_ENTRY_KEY').then(results => {
      expect(Object.keys(results[0])).toEqual([
        'word',
        'functional_label',
        'definition'
      ])
    })
  })
})

describe('walking <def> node', () => {
  it('creates new Sense instance when it encounter <sn>', () => {
    return dict.lookup('TEST_DEF_SN').then(results => {
      let { definition } = results[0]
      expect(definition).toHaveLength(2)
    })
  })
  it('creates new Sense instances when <def> node has no <sn> but <dt>', () => {
    return dict.lookup('TEST_DEF_DT').then(results => {
      let { definition } = results[0]
      expect(definition).toHaveLength(1)
    })
  })
  it('combines <dt>, <sd>, <set> when any of them appear together before next <sn>', () => {
    return dict.lookup('TEST_DEF_MULTIPLE_DT').then(results => {
      let { definition } = results[0]
      let sense = definition[0]
      expect(sense.meanings).toHaveLength(2)
    })
  })
})

describe('walking <dt> node', () => {
  it('collects <sx> into synonyms', () => {
    return dict.lookup('TEST_DT').then(results => {
      let { definition } = results[0]
      let sense = definition[0]
      expect(sense.synonyms).toHaveLength(2)
    })
  })
  it('collects <vi> into illustrations', () => {
    return dict.lookup('TEST_DT').then(results => {
      let { definition } = results[0]
      let sense = definition[1]
      expect(sense.illustrations).toHaveLength(1)
    })
  })
})

describe('error handling', () => {
  it('throws error when word could not be found', () => {
    return dict.lookup('TEST_INEXIST').catch(e => {
      expect(e).toBeInstanceOf(WordNotFoundError)
      expect(e.suggestions).toBeInstanceOf(Array)
    })
  })
})
