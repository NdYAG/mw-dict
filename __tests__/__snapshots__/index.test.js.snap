// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walking <def> node build hierarchy for word senses 1`] = `
Array [
  Sense {
    "meanings": Array [
      "SENSE 1",
    ],
    "number": "1",
    "transitivity": undefined,
  },
  Sense {
    "number": "2",
    "senses": Array [
      Sense {
        "meanings": Array [
          "SENSE 2 a",
        ],
        "number": "a",
        "transitivity": undefined,
      },
      Sense {
        "meanings": Array [
          "SENSE 2 b",
        ],
        "number": "b",
      },
    ],
  },
  Sense {
    "number": "3",
    "senses": Array [
      Sense {
        "number": "a",
        "senses": Array [
          Sense {
            "meanings": Array [
              "SENSE 3 a (1)",
            ],
            "number": "(1)",
            "transitivity": undefined,
          },
          Sense {
            "meanings": Array [
              "SENSE 3 a (2)",
            ],
            "number": "(2)",
          },
          Sense {
            "meanings": Array [
              "SENSE 3 a (3)",
            ],
            "number": "(3)",
          },
        ],
      },
      Sense {
        "number": "b",
        "senses": Array [
          Sense {
            "meanings": Array [
              "SENSE 3 b (1)",
            ],
            "number": "(1)",
          },
          Sense {
            "meanings": Array [
              "SENSE 3 b (2)",
            ],
            "number": "(2)",
          },
          Sense {
            "meanings": Array [
              "SENSE 3 b (3)",
            ],
            "number": "(3)",
          },
        ],
      },
      Sense {
        "meanings": Array [
          "SENSE 3 c",
        ],
        "number": "c",
      },
    ],
  },
]
`;
